keps/sig-auth/3766-referential-authorization/README.md

```diff
@@ -413,80 +422,115 @@ rejected".
+// pattern(s).
+type ClusterReferenceConsumer struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Subject refers to the subject that is a consumer of the referenced
+    // pattern(s).
+    Subject Subject       `json:"subject"`
+
+    // From refers to the group and resource that these references originate from.
+    From    GroupResource `json:"from"`
+
+    // To refers to the group and resource that these references target.
+    To      GroupResource `json:"to"`
+
+    // For refers to the purpose of this reference. (Cluster)ReferenceGrants
+    // matching the From, To, and For of this resource will be authorized for
+    // the Subject of this resource.
+    For     string        `json:"for"`
+}
 
-	// Spec defines the desired state of ReferenceGrant.
-	Spec ReferenceGrantSpec `json:"spec,omitempty"`
+// Subject is a copy of RBAC Subject that excludes APIGroup.
+type Subject struct {
+    // Kind of object being referenced. Values defined by this API group are
+    // "User", "Group", and "ServiceAccount". If the Authorizer does not
+    // recognized the kind value, the Authorizer should report an error.
+    Kind string `json:"kind"`
+    // Name of the object being referenced.
+    Name string `json:"name"`
+    // Namespace of the referenced object.  If the object kind is non-namespace,
+    // such as "User" or "Group", and this value is not empty the Authorizer
+    // should report an error. +optional
+    Namespace string `json:"namespace,omitempty"`
+}
+```
 
-	// Note that `Status` sub-resource has been excluded at the
-	// moment as it was difficult to work out the design.
-	// A `Status` sub-resource may be added in the future.
+#### ClusterReferenceGrant
+
+```golang
+// ClusterReferenceGrant authorizes same-namespace references matching the
+// specified pattern.
+type ClusterReferenceGrant struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // From refers to the group, resource, and pthat that these references
+    // originate from.
+    From    GroupVersionResourcePath `json:"from"`
+
+    // To refers to the group and resource that these references target.
+    To      GroupResource `json:"to"`
+
+    // For refers to the purpose of this reference. Subjects of
+    // ClusterReferenceConsumers will be authorized to follow references matching
+    // the From, To, and For of this resource.
+    For     string        `json:"for"`
 }
 
-// +kubebuilder:object:root=true
-// ReferenceGrantList contains a list of ReferenceGrant.
-type ReferenceGrantList struct {
-	metav1.TypeMeta `json:",inline"`
-	metav1.ListMeta `json:"metadata,omitempty"`
-	Items           []ReferenceGrant `json:"items"`
+type GroupVersionResourcePath struct {
+    Group    string `json:"group"`
+    Resource string `json:"resource"`
+    Version  string `json:"version"`
+    // Path is the path which this reference may come from.
```

Comment:
Shall we specify here that this needs to be a valid JSONPath?
And/or actually name it `jsonPath` to be explicit, and allow e.g. CEL to be neatly added in the future.

I guess this field is optional, then we should specify that and add omitempty.

Is there any risk that users would like to have a list of strings instead?
Of course, they could make two objects, but it's convenient to make this a list if that's deemed to be a common use-case here.

Reply:
Good call, I've renamed this to JSONPath. I think it's unlikely that a user would want to specify multiple JSONPath, but if they did, they could always create multiple ClusterReferenceGrants.